### PR TITLE
Migrate UDPRecievers to use Spring Kafka

### PR DESCRIPTION
**Pull request by mcook42**
# PR Details
## Description

- Added various missing and new Javadocs to increase code maintainability (and to appease CheckStyle)
- Introduced LoggingProducerListener for improved debugging experience
- Introduced DisabledTopicsProducerInterceptor to enable configuration driven topic
- Centralized Confluent property map construction to reduce duplication
- Replaced all hand-rolled kafka producers in the UDPReceivers flow with the Spring Kafka implementation
- Used values from the [udpsender files](https://github.com/CDOT-CV/jpo-ode/blob/0ceaa6bb0e9f5be761393e46fb87da395a69b1bd/scripts/tests) to create test data for the UDPReceivers
- Moved Producer-related code from `kafka` package to `kafka.producer` package
- Moved Test Utilities under `testUtilities` to `test.utilities` to follow convention

## Related Issue

No related USDOT issue

## Motivation and Context

Implementing [Spring Kafka ](https://docs.spring.io/spring-kafka/reference/index.html)gives us better lifecycle management of producers and consumers, more reusable producer/consumer code, easier testability, and a more robust production-ready Kafka library. This is part of a larger effort to replace our hand-rolled Kafka implementation with Spring Kafka. The previous changesets related to this effort are https://github.com/CDOT-CV/jpo-ode/pull/118, https://github.com/CDOT-CV/jpo-ode/pull/116, and https://github.com/CDOT-CV/jpo-ode/pull/123

Documentation I found helpful:
- https://docs.spring.io/spring-kafka/reference/kafka/interceptors.html
- https://docs.spring.io/spring-kafka/reference/kafka/producer-interceptor-managed-in-spring.html
- https://docs.spring.io/spring-kafka/reference/kafka/annotation-error-handling.html

## How Has This Been Tested?

Before any code was changed, various Unit and Integration Tests were added to the UDPReciever flows (all message types + GenericReceiver). These tests continued to pass after the code changes were made. I also added some tests for the new Spring Kafka components.

I started up the docker-compose locally with the profile set to `all` and ran the data in the receiver's respective `udpsender_[type].py` scripts through the UDP Receivers using Packet Sender. I confirmed in the logs that all messages passed through the expected topics.

I then added `"topic.OdeSpatJson"` to the disabled-topics list in application.yaml, restarted the service, and sent a SPaT message (see udpsender_spat.py for more detail) to confirm the message was not produced to `"topic.OdeRawEncodedSPATJson"` and the service handled the flow gracefully.

The logs look like the following (when debug logging is turned on)
```
2024-12-05 16:52:52 2024-12-05 23:52:52 [UdpReceiverManager] DEBUG SpatReceiver - Waiting for UDP SPaT packets... 
2024-12-05 16:52:52 2024-12-05 23:52:52 [UdpReceiverManager] DEBUG UdpHexDecoder - Packet received from 172.18.0.1:44153 
2024-12-05 16:52:52 2024-12-05 23:52:52 [UdpReceiverManager] DEBUG UdpHexDecoder - Full SPAT packet: 001338000817a780000089680500204642b342b34802021a15a955a940181190acd0acd20100868555c555c00104342aae2aae002821a155715570000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 
2024-12-05 16:52:52 2024-12-05 23:52:52 [UdpReceiverManager] DEBUG UperUtil - Base payload: 001338000817a780000089680500204642b342b34802021a15a955a940181190acd0acd20100868555c555c00104342aae2aae002821a155715570000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 
2024-12-05 16:52:52 2024-12-05 23:52:52 [UdpReceiverManager] DEBUG UperUtil - Stripped payload: 001338000817a780000089680500204642b342b34802021a15a955a940181190acd0acd20100868555c555c00104342aae2aae002821a15571557000 
2024-12-05 16:52:52 2024-12-05 23:52:52 [UdpReceiverManager] DEBUG UdpHexDecoder - Stripped SPAT packet: 001338000817a780000089680500204642b342b34802021a15a955a940181190acd0acd20100868555c555c00104342aae2aae002821a15571557000 
2024-12-05 16:52:52 2024-12-05 23:52:52 [UdpReceiverManager] WARN  SpatReceiver - Blocked attempt to send data to disabled topic topic.OdeRawEncodedSPATJson
```

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[ODE Contributing Guide](https://github.com/usdot-jpo-ode/jpo-ode/blob/bugfix/Pull_request_template/docs/contributing_guide.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/bsm/BsmReceiverTest.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+class BsmReceiverTest {
+
+  @Autowired
+  UDPReceiverProperties udpReceiverProperties;
+
+  @Autowired
+  RawEncodedJsonTopics rawEncodedJsonTopics;
+
+  @Autowired
+  KafkaTemplate<String, String> kafkaTemplate;
```
Comment (dmccoystephenson) [inline jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/bsm/BsmReceiverTest.java:58]: (praise) I didn't realize that you could use `@Autowired` in test classes; that's cool!

File: jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/bsm/BsmReceiverTest.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+  EmbeddedKafkaBroker embeddedKafka = EmbeddedKafkaHolder.getEmbeddedKafka();
+
+  @Test
+  void testRun() throws Exception {
+    // create the needed topic for production
+    try {
+      embeddedKafka.addTopics(new NewTopic(rawEncodedJsonTopics.getBsm(), 1, (short) 1));
+    } catch (Exception e) {
+      // ignore because we only care that the topics exist not that they're unique
```
Comment (dmccoystephenson) [inline jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/bsm/BsmReceiverTest.java:68]: (non-blocking) This comment seems to indicate that duplicate topics can exist; would this ever be a problem for testing?

File: jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/generic/GenericReceiverTest.java
Comment (dmccoystephenson) [inline jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/generic/GenericReceiverTest.java:1]: (blocking) It doesn't look like the SRM path is being tested in this class, is that something we want to do here?

File: jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/spat/SpatReceiverTest.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    JSONObject producedJson = new JSONObject(singleRecord.value());
+    JSONObject expectedJson = new JSONObject(expected);
+
+    assertNotEquals(expectedJson.getJSONObject("metadata").get("serialId"),
+        producedJson.getJSONObject("metadata").get("serialId"));
+    expectedJson.getJSONObject("metadata").remove("serialId");
+    producedJson.getJSONObject("metadata").remove("serialId");
+
+    assertEquals(
+        expectedJson.toString(2),
```
Comment (dmccoystephenson) [inline jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/spat/SpatReceiverTest.java:107]: (non-blocking) What does passing '2' to `toString()` do for us here?

**Review by dmccoystephenson — CHANGES_REQUESTED** [review]
This looks good to me for the most part! The unit tests passed when I ran them locally. I just had some non-blocking comments (mostly questions) and a single blocking comment.

File: jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/generic/GenericReceiverTest.java
Comment (mcook42) [reply jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/generic/GenericReceiverTest.java:1]: yes, nice catch. I've added that here https://github.com/CDOT-CV/jpo-ode/pull/129/commits/79c411ca267f8de7bcd8390d7396acde24045929

File: jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/spat/SpatReceiverTest.java (hunk 3 above)
Comment (mcook42) [reply jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/spat/SpatReceiverTest.java:107]: It sets the indentation on the stringified JSON to two spaces & indicates that the JSON should be pretty-printed. Comparing pretty-printed JSON makes it easier for diff tools to identify which part of the JSON is mismatched. Without the pretty-print, the JSON is all on one line so the diff tools have a hard time highlighting specific JSON elements

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/controller/UdpServicesController.java [hunk 4]
```diff
@@ -22,22 +21,31 @@
...
+   */
+  @Autowired
+  public UdpServicesController(UDPReceiverProperties udpProps,
+      RawEncodedJsonTopics rawEncodedJsonTopics,
+      KafkaTemplate<String, String> kafkaTemplate) {
+    super();
 
-        ServiceManager rm = new ServiceManager(new UdpServiceThreadFactory("UdpReceiverManager"));
-        log.debug("Starting UDP receiver services...");
+    ServiceManager rm = new ServiceManager(new UdpServiceThreadFactory("UdpReceiverManager"));
```
[the remark this answers is not in the document]
Comment (mcook42) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/controller/UdpServicesController.java]: good callout. updated here https://github.com/CDOT-CV/jpo-ode/pull/129/commits/7954b4a1d740bf833e25a8de017cc1a82d8694db

File: jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/bsm/BsmReceiverTest.java (hunk 2 above)
Comment (dmccoystephenson) [reply jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/bsm/BsmReceiverTest.java:68]: Sounds good!

File: jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/spat/SpatReceiverTest.java (hunk 3 above)
Comment (dmccoystephenson) [reply jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/udp/spat/SpatReceiverTest.java:107]: That's awesome!

**Review by dmccoystephenson — APPROVED** [review]
Looks good & the unit tests pass!

**Review by drewjj — APPROVED** [review]
All message feeds are functional and the updated tests maker sense and are functional. Nice job!

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/generic/GenericReceiver.java [hunk 5]
```diff
@@ -1,111 +1,140 @@
...
+/**
+ * GenericReceiver is a class that listens for UDP packets and processes them based on the
+ * determined message type. It extends AbstractUdpReceiverPublisher to take advantage of the
+ * runnable interface for running the receiver service in a separate thread.
+ *
+ * </p>The class is designed to handle all {@link us.dot.its.jpo.ode.uper.SupportedMessageType}
+ * message types encoded in UDP packets such as and routes them to the appropriate Kafka topic.
+ */
 @Slf4j
 public class GenericReceiver extends AbstractUdpReceiverPublisher {
```
[the remark this answers is not in the document]
Comment (dmccoystephenson) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/generic/GenericReceiver.java:25]: Sounds good, thanks John!

**Pull request merged**